### ISSUE TITLE
Explainability and interpretability in binary text categorization

### ISSUE THREAD

**Issue by rth** [FreeDiscovery/FreeDiscovery#32]
Unlike  LSI, where a document is categorized  based on it's cosine distance to the nearest relevant and non relevant document (thus inherently being explainable), ML based categorization is more difficult to explain.

This issue thus aims to track all development relevant to explainability and interpretabilty of ML  binary text categorization models.

Two independent approaches could be,
  1. Given a trained ML model, illustrate which n-grams (or word) have an influence on it being categorized as relevant / non relevant (issue #33)
  2. Given a list of documents ids (could false positives, false negatives), display some statistics that might help understanding why they were incorrectly categorized (issue #)  

## References

  -  [Accuracy vs Explainability of Machine Learning Models](http://www.inference.vc/accuracy-vs-explainability-in-machine-learning-models-nips-workshop-poster-review/)
 - R. Turner et al (2015): [A Model Explanation System](http://www.blackboxworkshop.org/pdf/Turner2015_MES.pdf)
  - A. Vellido et al (2014): [Making machine learning models interpretable](https://pdfs.semanticscholar.org/ce0b/8b6fca7dc089548cc2e9aaac3bae82bb19da.pdf)
  - Martens et al (2007): [Comprehensible credit scoring models using rule extraction from support vector machines](http://www.sciencedirect.com/science/article/pii/S0377221706011878)

**Comment by DenisFLASH** [comment]
Wow, looks like "failure analysis" which we were talking about, is quite a complex subject, with several referenced research works and so on..
In which version you want to include the resolution of this issue?

**Comment by rth** [comment]
Actually, the implementation in [ELI5](https://github.com/TeamHG-Memex/eli5), based on the [LIME]( http://eli5.readthedocs.io/en/latest/lime.html) algorithm (Ribeiro et al., 2016) should allow to also handle non linear models.

**Comment by rth** [comment]
A base implementation was done in #33, anything more advanced should probably use eli5.
